ksqldb-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java

```diff
@@ -293,6 +319,195 @@ public TransientQueryMetadata executeTransientQuery(
+    // Using a single thread keeps these queries as lightweight as possible, since we are
+    // not counting them against the transient query limit.
+    // Setting the commit interval to 100ms to make the query results snappier.
+    // Since this is a "sit and wait", not a "fire and forget", query,
+    // we want to make sure people are going to see their results come back asap.
+    final ConfiguredStatement<Query> statement = statementOrig.withConfigOverrides(
+        ImmutableMap.<String, Object>builder()
+            .putAll(statementOrig.getSessionConfig().getOverrides())
+            .put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+            .put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 1)
+            .put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100)
+            .build()
+    );
+    try {
+
+      final TransientQueryMetadata transientQueryMetadata = EngineExecutor
+          .create(primaryContext, serviceContext, statement.getSessionConfig())
+          .executeTransientQuery(statement, excludeTombstones);
+
+      QueryLogger.info(
+          "Streaming stream pull query results '{}'",
+          statement.getStatementText()
+      );
+
+      return transientQueryMetadata;
+
+    } catch (final KsqlStatementException e) {
+      throw e;
+    } catch (final KsqlException e) {
+      // add the statement text to the KsqlException
+      throw new KsqlStatementException(e.getMessage(), statement.getStatementText(), e.getCause());
+    }
+  }
+
+  public void waitForStreamPullQuery(
+      final ServiceContext serviceContext,
+      final ImmutableAnalysis analysis,
+      final ConfiguredStatement<Query> statement,
+      final TransientQueryMetadata transientQueryMetadata) {
+    // query the endOffsets of the input
+    final String sourceTopicName = analysis.getFrom().getDataSource().getKafkaTopicName();
+    final Admin admin = serviceContext.getAdminClient();
+    final TopicDescription topicDescription = getTopicDescription(
+        admin,
+        sourceTopicName
+    );
+
+    final Object processingGuarantee = statement
+        .getSessionConfig()
+        .getConfig(true)
+        .getKsqlStreamConfigProps()
+        .get(StreamsConfig.PROCESSING_GUARANTEE_CONFIG);
+
+    final IsolationLevel isolationLevel =
+        StreamsConfig.AT_LEAST_ONCE.equals(processingGuarantee)
+            ? IsolationLevel.READ_UNCOMMITTED
+            : IsolationLevel.READ_COMMITTED;
+
+    final Map<TopicPartition, Long> endOffsets = getEndOffsets(
+        admin,
+        topicDescription,
+        isolationLevel
+    );
+
+    // wait for the query to pass the endOffsets
+    while (!passedEndOffsets(admin, transientQueryMetadata, endOffsets)) {
+      if (transientQueryMetadata.getState() == State.ERROR) {
+        log.error("Stream pull query execution has failed");
+        throw new KsqlServerException("Server error");
+      } else {
+        try {
+          Thread.sleep(50L);
```

Comment:
One thing that occurs to me is that for the async endpoints, you're not going to be able to call this directly on the request thread.  For the reasons I mentioned elsewhere, I think you're going to have to do this polling in a background thread anyway to match the current invariants, so that seems like a reasonable solution.